dnf/cli/commands/repoquery.py

```diff
@@ -340,55 +346,35 @@ def build_format_fn(self, opts, pkg):
-        t = self.base.sack.query().filterm(empty=True)
-        set_requires = set()
-        set_all_deps = set()
-
-        for pkg in query_select.run():
-            pkg_provides = pkg.provides
-            set_requires.update(pkg_provides)
-            set_requires.update(pkg.files)
-            if all_deps:
-                set_all_deps.update(pkg_provides)
-
-        t = t.union(query_in.filter(requires=set_requires))
-        if set_all_deps:
-            t = t.union(query_in.filter(recommends=set_all_deps))
-            t = t.union(query_in.filter(enhances=set_all_deps))
-            t = t.union(query_in.filter(supplements=set_all_deps))
-            t = t.union(query_in.filter(suggests=set_all_deps))
-        if recursive:
-            query_select = t.difference(done)
-            if query_select:
-                done = self._get_recursive_deps_query(query_in, query_select, done=t.union(done),
-                                                      recursive=recursive, all_deps=all_deps)
-        return t.union(done)
+    def _do_recursive_deps(self, query_in, query_select, done=None):
+        done = done if done else query_select
+
+        query_required = query_in.filter(requires=query_select, match_provides=True)
+
+        query_select = query_required.difference(done)
+        done = query_required.union(done)
+
+        if query_select:
+            done = self._do_recursive_deps(query_in, query_select, done=done)
 
-    def by_all_deps(self, requires_name, depends_name, query):
-        names = requires_name or depends_name
-        defaultquery = self.base.sack.query().filterm(empty=True)
-        for name in names:
-            defaultquery = defaultquery.union(query.intersection(
-                dnf.subject.Subject(name).get_best_query(self.base.sack, with_provides=False,
-                                                         with_filenames=False)))
-        requiresquery = query.filter(requires__glob=names)
-        if depends_name:
-            requiresquery = requiresquery.union(query.filter(recommends__glob=depends_name))
-            requiresquery = requiresquery.union(query.filter(enhances__glob=depends_name))
-            requiresquery = requiresquery.union(query.filter(supplements__glob=depends_name))
-            requiresquery = requiresquery.union(query.filter(suggests__glob=depends_name))
-
-        done = requiresquery.union(self._get_recursive_deps_query(query, defaultquery,
-                                                                  all_deps=depends_name))
-        if self.opts.recursive:
-            done = done.union(self._get_recursive_deps_query(query, done,
-                                                             recursive=self.opts.recursive,
-                                                             all_deps=depends_name))
         return done
 
+    def by_all_deps(self, names, query, all_dep_types=False):
+        depquery = query.filter(requires__glob=names, match_provides=True)
```

Comment:
It looks like that it drops support of NEVRA argument for `--whatrequires=dnf-2.7.5`. I did a small survey and I discovered that 100% of user uses as a argument NEVRA for the option and 50% of user never used provides for it. I would considered that such a change cannot land in dnf4, and I am also against for dnf5.

Reply:
Haha, ok. But you should also state your sample size to make your data a bit more meaningful :wink: (I hope it weren't two people)

In that case, we need to think of how to support NEVRAs in these options explicitly and consistently. (instead of by accident...)